arangod/Network/Utils.cpp

```diff
@@ -0,0 +1,226 @@
+    serverID = dest.substr(7);
+  } else {
+    std::string errorMessage = "did not understand destination'" + dest + "'";
+    LOG_TOPIC(ERR, Logger::COMMUNICATION)
+        << "did not understand destination '" << dest << "'";
+    return TRI_ERROR_CLUSTER_BACKEND_UNAVAILABLE;
+  }
+
+  endpoint = ci->getServerEndpoint(serverID);
+  if (endpoint.empty()) {
+    if (serverID.find(',') != std::string::npos) {
+      TRI_ASSERT(false);
+    }
+    std::string errorMessage =
+        "did not find endpoint of server '" + serverID + "'";
+    LOG_TOPIC(ERR, Logger::COMMUNICATION)
+        << "did not find endpoint of server '" << serverID << "'";
+    return TRI_ERROR_CLUSTER_BACKEND_UNAVAILABLE;
+  }
+  return TRI_ERROR_NO_ERROR;
+}
+
+OperationResult errorFromBody(arangodb::velocypack::Buffer<uint8_t> const& body,
+                              int defaultErrorCode) {
+  if (body.size() > 0) {
+    return errorFromBody(VPackSlice(body.data()), defaultErrorCode);
+  }
+  return OperationResult(defaultErrorCode);
+}
+
+OperationResult errorFromBody(std::shared_ptr<VPackBuilder> const& body, int defaultErrorCode) {
+  if (body) {
+    return errorFromBody(body->slice(), defaultErrorCode);
+  }
+  return OperationResult(defaultErrorCode);
+}
+
+OperationResult errorFromBody(VPackSlice const& body, int defaultErrorCode) {
+  // read the error number from the response and use it if present
+  if (body.isObject()) {
+    VPackSlice num = body.get(StaticStrings::ErrorNum);
+    VPackSlice msg = body.get(StaticStrings::ErrorMessage);
+    if (num.isNumber()) {
+      if (msg.isString()) {
+        // found an error number and an error message, so let's use it!
+        return OperationResult(Result(num.getNumericValue<int>(), msg.copyString()));
+      }
+      // we found an error number, so let's use it!
+      return OperationResult(num.getNumericValue<int>());
+    }
+  }
+
+  return OperationResult(defaultErrorCode);
+}
+
+/// @brief extract the error code form the body
+int errorCodeFromBody(arangodb::velocypack::Slice const& body) {
+  if (body.isObject()) {
+    VPackSlice num = body.get(StaticStrings::ErrorNum);
+    if (num.isNumber()) {
+      // we found an error number, so let's use it!
+      return num.getNumericValue<int>();
+    }
+  }
+  return TRI_ERROR_ILLEGAL_NUMBER;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief Extract all error baby-style error codes and store them in a map
+////////////////////////////////////////////////////////////////////////////////
+
+void errorCodesFromHeaders(network::Headers headers,
+                           std::unordered_map<int, size_t>& errorCounter,
+                           bool includeNotFound) {
+  auto const& codes = headers.find(StaticStrings::ErrorCodes);
+  if (codes != headers.end()) {
+    auto parsedCodes = VPackParser::fromJson(codes->second);
+    VPackSlice codesSlice = parsedCodes->slice();
+    TRI_ASSERT(codesSlice.isObject());
```

Comment:
I think we shouldn't assert on this, because some attacker can use this to crash the server (though only in maintainer mode)